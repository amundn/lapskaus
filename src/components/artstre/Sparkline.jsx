import React from 'react'
import { Sparklines, SparklinesLine, SparklinesBars } from 'react-sparklines'

class SparkLineContainer extends React.Component {
  constructor() {
    super()
    this.state = { data: [] }
  }
  componentWillUpdate(nextProps, nextState) {
        if (nextProps.taxonId !== this.props.taxonId)
        this.loadData(nextProps.taxonId)
  }

  loadData(taxonId) {
    const url = 'http://localhost:8888/stat/' + taxonId + '.json'
    console.log(url)
//    this.setState({ data: [] })
    fetch(url)
      .then(response => response.json())
      .then(json => this.setState({ data: json.distribution }))
      .catch(err => console.error(err))
  }

  componentWillMount() {
    this.loadData(this.props.taxonId)
  }

  render() {
    return (
        <SparkLine data={this.state.data} {...this.props} />
    )
  }
}

const SparkLine = ({ taxonId, data, color }) => {
  return (
    <div style={{ width: 200 }}>
      <Sparklines data={data} width={100} height={20} margin={2}>
        <SparklinesLine color={color} />
        <SparklinesBars style={{ fill: 'slategray', fillOpacity: '.5' }} />
      </Sparklines>
{false &&      <div>{taxonId}_{data.join(',')}</div>}
    </div>
  )
}

export default SparkLineContainer
