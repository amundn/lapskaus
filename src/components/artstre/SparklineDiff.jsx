import React from 'react'
import {
  Sparklines,
  SparklinesNormalBand,
  SparklinesReferenceLine,
  SparklinesLine,
  SparklinesBars
} from 'react-sparklines'
import config from '../../config'

class SparkLineContainer extends React.Component {
  constructor() {
    super()
    this.state = { data: [] }
  }
  componentWillUpdate(nextProps, nextState) {
    if (nextProps.taxonId !== this.props.taxonId)
      this.loadData(nextProps.taxonId, 'data')
    if (nextProps.taxonIdParent !== this.props.taxonIdParent)
      this.loadData(nextProps.taxonIdParent, 'parentData')
  }

  loadData(taxonId, target) {
    const url = config.apiUrl + 'stat/' + taxonId + '.json'

    console.error('===')
    this.context.fetchJson('Statistikk', url, json => {
      var ns = {}
      ns[target] = json.distribution
      this.setState(ns)
    })
    console.error('===')
    /*

    fetch(url)
      .then(response => response.json())
      .then(json => {
        var ns = {}
        ns[target] = json.distribution
        this.setState(ns)
      })
      .catch(err => console.error(err))*/
  }

  componentWillMount() {
    this.loadData(this.props.taxonId, 'data')
    this.loadData(this.props.taxonIdParent, 'parentData')
  }

  render() {
    if (!this.state.data || !this.state.parentData) return null
    const data = this.calculate(this.state.data, this.state.parentData)
    return <SparkLine data={data} {...this.props} />
  }

  calculate(child, parent) {
    const p = this.normalize(parent)
    const c = this.normalize(child)
    //    for (var i = 0; i < c.length; i++) c[i] = (c[i] - p[i]) / p[i]
    for (var i = 0; i < c.length; i++) c[i] = c[i] - p[i]

    return c
  }

  normalize(dist) {
    let count = 0
    let r = dist.slice(0)
    for (var i = 0; i < dist.length; i++) count += r[i]
    for (i = 0; i < dist.length; i++) r[i] /= count
    return r
  }
}

const SparkLine = ({ data, color }) => (
  <div style={{ width: 200 }}>
    <Sparklines data={data} width={100} height={30} margin={2}>
      <SparklinesLine color={color} />
      <SparklinesReferenceLine type="mean" />
      <SparklinesBars style={{ fill: 'slategray', fillOpacity: '.5' }} />
      <SparklinesNormalBand />
    </Sparklines>
  </div>
)

export default SparkLineContainer
