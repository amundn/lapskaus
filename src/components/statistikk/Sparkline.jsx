import React from 'react'
import { Sparklines, SparklinesLine, SparklinesBars } from 'react-sparklines'

const SparkLine = ({ taxonId, data, color }) => {
  return (
    <div style={{ width: 200 }}>
      <Sparklines data={data} width={100} height={20} margin={2}>
        <SparklinesLine color={color} />
        <SparklinesBars style={{ fill: 'slategray', fillOpacity: '.5' }} />
      </Sparklines>
{false &&      <div>{taxonId}_{data.join(',')}</div>}
    </div>
  )
}

export default SparkLine
