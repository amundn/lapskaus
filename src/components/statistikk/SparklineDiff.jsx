import React from 'react'
import {
  Sparklines,
  SparklinesNormalBand,
  SparklinesReferenceLine,
  SparklinesLine,
  SparklinesBars
} from 'react-sparklines'

const SparkLineDiff = ({ data, color }) => (
  <div style={{ width: 200 }}>
    <Sparklines data={data} width={100} height={30} margin={2}>
      <SparklinesLine color={color} />
      <SparklinesReferenceLine type="mean" />
      <SparklinesBars style={{ fill: 'slategray', fillOpacity: '.5' }} />
      <SparklinesNormalBand />
    </Sparklines>
  </div>
)

export default SparkLineDiff
