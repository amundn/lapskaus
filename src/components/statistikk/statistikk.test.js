import React from 'react'
import renderer from 'react-test-renderer'
import Sparkline from './Sparkline'
import SparklineDiff from './SparklineDiff'

test('Sparkline', () => {
  const component = renderer.create(<Sparkline color='red' data={[1, 2, 3]} taxonId={4} />)
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('Sparkline diff', () => {
  const component = renderer.create(<SparklineDiff color='red' data={[1, 2, 3]} taxonId={4} />)
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})
