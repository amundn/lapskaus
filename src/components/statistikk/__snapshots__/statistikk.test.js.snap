// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sparkline 1`] = `
<div
  style={
    Object {
      "width": 200,
    }
  }
>
  <svg
    preserveAspectRatio="none"
    style={undefined}
    viewBox="0 0 100 20"
  >
    <g>
      <circle
        cx={2}
        cy={18}
        onClick={[Function]}
        onMouseEnter={[Function]}
        r={2}
        style={
          Object {
            "fill": "red",
            "fillOpacity": ".1",
            "pointerEvents": "auto",
            "stroke": "none",
            "strokeWidth": "0",
          }
        }
      />
      <circle
        cx={50}
        cy={10}
        onClick={[Function]}
        onMouseEnter={[Function]}
        r={2}
        style={
          Object {
            "fill": "red",
            "fillOpacity": ".1",
            "pointerEvents": "auto",
            "stroke": "none",
            "strokeWidth": "0",
          }
        }
      />
      <circle
        cx={98}
        cy={2}
        onClick={[Function]}
        onMouseEnter={[Function]}
        r={2}
        style={
          Object {
            "fill": "red",
            "fillOpacity": ".1",
            "pointerEvents": "auto",
            "stroke": "none",
            "strokeWidth": "0",
          }
        }
      />
      <polyline
        points="2 18 50 10 98 2 98 18 2 18 2 18"
        style={
          Object {
            "fill": "red",
            "fillOpacity": ".1",
            "pointerEvents": "auto",
            "stroke": "none",
            "strokeWidth": "0",
          }
        }
      />
      <polyline
        points="2 18 50 10 98 2"
        style={
          Object {
            "fill": "none",
            "stroke": "red",
            "strokeLinecap": "round",
            "strokeLinejoin": "round",
            "strokeWidth": "1",
          }
        }
      />
    </g>
    <g
      transform="scale(1,-1)"
    >
      <rect
        height={2}
        onMouseMove={undefined}
        style={
          Object {
            "fill": "slategray",
            "fillOpacity": ".5",
          }
        }
        width={44}
        x={-20}
        y={-20}
      />
      <rect
        height={10}
        onMouseMove={undefined}
        style={
          Object {
            "fill": "slategray",
            "fillOpacity": ".5",
          }
        }
        width={44}
        x={28}
        y={-20}
      />
      <rect
        height={18}
        onMouseMove={undefined}
        style={
          Object {
            "fill": "slategray",
            "fillOpacity": ".5",
          }
        }
        width={44}
        x={76}
        y={-20}
      />
    </g>
  </svg>
</div>
`;

exports[`Sparkline diff 1`] = `
<div
  style={
    Object {
      "width": 200,
    }
  }
>
  <svg
    preserveAspectRatio="none"
    style={undefined}
    viewBox="0 0 100 30"
  >
    <g>
      <circle
        cx={2}
        cy={28}
        onClick={[Function]}
        onMouseEnter={[Function]}
        r={2}
        style={
          Object {
            "fill": "red",
            "fillOpacity": ".1",
            "pointerEvents": "auto",
            "stroke": "none",
            "strokeWidth": "0",
          }
        }
      />
      <circle
        cx={50}
        cy={15}
        onClick={[Function]}
        onMouseEnter={[Function]}
        r={2}
        style={
          Object {
            "fill": "red",
            "fillOpacity": ".1",
            "pointerEvents": "auto",
            "stroke": "none",
            "strokeWidth": "0",
          }
        }
      />
      <circle
        cx={98}
        cy={2}
        onClick={[Function]}
        onMouseEnter={[Function]}
        r={2}
        style={
          Object {
            "fill": "red",
            "fillOpacity": ".1",
            "pointerEvents": "auto",
            "stroke": "none",
            "strokeWidth": "0",
          }
        }
      />
      <polyline
        points="2 28 50 15 98 2 98 28 2 28 2 28"
        style={
          Object {
            "fill": "red",
            "fillOpacity": ".1",
            "pointerEvents": "auto",
            "stroke": "none",
            "strokeWidth": "0",
          }
        }
      />
      <polyline
        points="2 28 50 15 98 2"
        style={
          Object {
            "fill": "none",
            "stroke": "red",
            "strokeLinecap": "round",
            "strokeLinejoin": "round",
            "strokeWidth": "1",
          }
        }
      />
    </g>
    <line
      style={
        Object {
          "stroke": "red",
          "strokeDasharray": "2, 2",
          "strokeOpacity": 0.75,
        }
      }
      x1={2}
      x2={98}
      y1={17}
      y2={17}
    />
    <g
      transform="scale(1,-1)"
    >
      <rect
        height={2}
        onMouseMove={undefined}
        style={
          Object {
            "fill": "slategray",
            "fillOpacity": ".5",
          }
        }
        width={44}
        x={-20}
        y={-30}
      />
      <rect
        height={15}
        onMouseMove={undefined}
        style={
          Object {
            "fill": "slategray",
            "fillOpacity": ".5",
          }
        }
        width={44}
        x={28}
        y={-30}
      />
      <rect
        height={28}
        onMouseMove={undefined}
        style={
          Object {
            "fill": "slategray",
            "fillOpacity": ".5",
          }
        }
        width={44}
        x={76}
        y={-30}
      />
    </g>
    <rect
      height={NaN}
      style={
        Object {
          "fill": "red",
          "fillOpacity": 0.1,
        }
      }
      width={96}
      x={2}
      y={6.385544447939562}
    />
  </svg>
</div>
`;
